### ISSUE TITLE
Performance via Chunks

### ISSUE THREAD

**Issue by lumi-a** [Tim2othy/spacegame#11]
Currently, a lot of objects interact pairwise:
- Each planet interects with each {Asteroid, Ship}
- Each bullet interacts with each {Planet, Asteroid, some ships}

If we eventually expand the map to be much larger, containing many more objects, these interactions will kill performance. A solution to this:

Chunks are squares that tile the whole universe. Each object is assigned to a chunk, and when it moves outside of its current chunk, it gets assigned to a new one.

An object `A` interacts (gravity, intersection-checks) only with objects `B` that are in "close" chunks. In pseudocode, this might mean that, instead of the current interactions:
```
for A in planents:
    for b in ships:
        a.interact(b)
```
we would have something like:
```
ADJACENT = [(-1,-1),(0,-1),(1,-1), (-1,0),(0,0),(1,0), (-1,1),(0,1),(1,1)]
for A in planets:
    adjacent_chunks = [A.chunk + v for v in ADJACENT]
    for b in get_ships_in_chunks(adjacent_chunks):
        a.interact(b)
```
The size of the chunks would have to be chosen appropriately, of course.

Chunks would change the way gravity acts: Originally, gravity does always act between two objects, no matter how far apart. With chunks, it'd only act between two objects that aren't too far apart. The effect should be negligible, though, because gravity is proportional to the inverse of squared distance.

**Comment by lumi-a** [comment]
Yes. In fact, we'd only calculate the square of the distance, and check if that is small (the same trick is already implemented in bouncing). Calculating the square of the distance is faster, as it doesn't involve a sqrt-operation. When gravitational-force is calculated in the end, we do need the distance, though.

**Comment by lumi-a** [comment]
Correcting my above comment: Just checking the square of the distance is not sufficient, the mass should also be a factor. Otherwise, in the extreme case, we could end up with huge planets that exert zero effective gravity.

**Comment by lumi-a** [comment]
If we do decide to implement chunks (which seems more attractive now with #56), it might even be possible to implement "infinite" universes, with similar techniques for generating enemies, asteroids, planets as described for the star-generation in #56, with little additional overhead. I don't know what your vision for the game is, whether an infinite universe makes sense.

**Comment by Tim2othy** [comment]
I tried to give you my vision here https://github.com/Tim2othy/spacegame/issues/63#issue-2857708322. Depending on how the AIs are implemented an infinite universe might make that harder, but it's hard to tell.
